datafusion/core/src/physical_plan/row_converter.rs

```diff
@@ -0,0 +1,292 @@
+#[derive(Debug)]
+pub enum CardinalityAwareRowConverter {
+    /// Converter is newly initialized, and hasn't yet seen data
+    New {
+        /// Defines the Row conversion
+        fields: Vec<SortField>,
+    },
+    /// Converter has seen data and can convert [`Array`]s to/from
+    /// [`Rows`]
+    Converting {
+        /// Underlying converter
+        converter: RowConverter,
+        /// if preserve_dictionaries is disabled, the output type can be
+        /// different than input.
+        output_types: Vec<DataType>,
+    },
+}
+
+impl CardinalityAwareRowConverter {
+    pub fn new(fields: Vec<SortField>) -> Result<Self> {
+        Ok(Self::New { fields })
+    }
+
+    /// Returns the memory size of the underlying [`RowConverter`] if
+    /// any.
+    pub fn size(&self) -> usize {
+        match self {
+            Self::New { .. } => 0,
+            Self::Converting { converter, .. } => converter.size(),
+        }
+    }
+
+    pub fn empty_rows(&self, row_capacity: usize, data_capacity: usize) -> Result<Rows> {
+        match self {
+            Self::New { .. } => internal_err!(
+                "CardinalityAwareRowConverter has not converted any rows yet"
+            ),
+            Self::Converting { converter, .. } => {
+                Ok(converter.empty_rows(row_capacity, data_capacity))
+            }
+        }
+    }
+
+    pub fn convert_rows<'a, I>(&self, rows: I) -> Result<Vec<ArrayRef>>
+    where
+        I: IntoIterator<Item = Row<'a>>,
+    {
+        match self {
+            Self::New { .. } => internal_err!(
+                "CardinalityAwareRowConverter has not converted any rows yet"
+            ),
+            Self::Converting {
+                converter,
+                output_types,
+            } => {
+                // Cast output type if needed
+                let output = converter
+                    .convert_rows(rows)?
+                    .into_iter()
+                    .zip(output_types.iter())
+                    .map(|(arr, output_type)| {
+                        if arr.data_type() != output_type {
+                            Ok(arrow::compute::cast(&arr, output_type)?)
+                        } else {
+                            Ok(arr)
+                        }
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                Ok(output)
+            }
+        }
+    }
+
+    /// Calls [`RowConverter::convert_columns`] after first
+    /// initializing the converter based on cardinalities
+    pub fn convert_columns(&mut self, columns: &[ArrayRef]) -> Result<Rows> {
+        Ok(self.converter_mut(columns)?.convert_columns(columns)?)
+    }
+
+    /// Return a mutable reference to the inner converter, creating it if needed
+    fn converter_mut(&mut self, columns: &[ArrayRef]) -> Result<&mut RowConverter> {
+        if let Self::New { fields } = self {
+            // TODO clean up the code
+            let mut updated_fields = fields.clone();
```

Comment:
Could use std::mem::take here to avoid a clone

Reply:
Sadly it seems like `SortOptions` does not implement `Default` 😢 


```
   --> datafusion/core/src/physical_plan/row_converter.rs:150:53
    |
150 |             let mut updated_fields = std::mem::take(&mut fields);
    |                                      -------------- ^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `&mut std::vec::Vec<arrow::arrow_row::SortField>`
    |                                      |
    |                                      required by a bound introduced by this call
    |
note: required by a bound in `std::mem::take`
```